File: python/analyzers/code_analyzer.py
"""
Unified Code Analyzer
Ported from SpineHUB.

Orchestrates multiple code analysis tools and provides
a unified interface for code quality checking.
"""

import json
import subprocess
import sys
import re
from pathlib import Path
from typing import Optional
from dataclasses import dataclass

from .analyzer_base import AnalyzerBase, AnalyzerResult, Issue, Severity


def run_python_module(module: str, args: list[str], cwd: str = None) -> subprocess.CompletedProcess:
    """Run a Python module via python -m for better Windows compatibility."""
    return subprocess.run(
        [sys.executable, "-m", module] + args,
        capture_output=True,
        text=True,
        cwd=cwd,
    )


class RuffAnalyzer(AnalyzerBase):
    """
    Ruff - Fast Python linter and formatter.

    Checks for:
    - Style violations (E, W codes)
    - Import issues (I codes)
    - Pyflakes errors (F codes)
    - And many more
    """

    name = "ruff"
    description = "Fast Python linter (10-100x faster than flake8)"

    def is_available(self) -> bool:
        try:
            result = run_python_module("ruff", ["--version"])
            return result.returncode == 0
        except Exception:
            return False

    def run(self, paths: Optional[list[str]] = None) -> AnalyzerResult:
        if not self.is_available():
            return AnalyzerResult(
                tool=self.name,
                success=False,
                error="Ruff is not installed. Run: pip install ruff",
            )

        target = paths if paths else [str(self.project_path)]

        try:
            result = run_python_module(
                "ruff",
                ["check", "--output-format=json"] + target,
                cwd=str(self.project_path),
            )

            issues = self.parse_output(result.stdout)

            return AnalyzerResult(
                tool=self.name,
                success=True,
                issues=issues,
                raw_output=result.stdout,
                summary={
                    "total_issues": len(issues),
                    "fixable": sum(1 for i in issues if i.fix_available),
                },
            )
        except Exception as e:
            return AnalyzerResult(
                tool=self.name,
                success=False,
                error=str(e),
            )

    def parse_output(self, output: str) -> list[Issue]:
        if not output.strip():
            return []

        try:
            data = json.loads(output)
        except json.JSONDecodeError:
            return []

        issues = []
        for item in data:
            # Determine severity based on code
            code = item.get("code", "")
            if code.startswith("E"):
                severity = Severity.ERROR
            elif code.startswith("W"):
                severity = Severity.WARNING
            else:
                severity = Severity.INFO

            issues.append(
                Issue(
                    file=item.get("filename", ""),
                    line=item.get("location", {}).get("row", 0),
                    column=item.get("location", {}).get("column", 0),
                    code=code,
                    message=item.get("message", ""),
                    severity=severity,
                    tool=self.name,
                    fix_available=item.get("fix") is not None,
                    fix_description=item.get("fix", {}).get("message") if item.get("fix") else None,
                )
            )

        return issues


class BanditAnalyzer(AnalyzerBase):
    """
    Bandit - Security-focused Python linter.

    Detects:
    - SQL injection (B608)
    - Command injection (B602, B603)
    - Hardcoded passwords (B105, B106)
    - Unsafe deserialization (B301, B302)
    """

    name = "bandit"
    description = "Python security scanner"

    def is_available(self) -> bool:
        try:
            result = run_python_module("bandit", ["--version"])
            return result.returncode == 0
        except Exception:
            return False

    def run(self, paths: Optional[list[str]] = None) -> AnalyzerResult:
        if not self.is_available():
            return AnalyzerResult(
                tool=self.name,
                success=False,
                error="Bandit is not installed. Run: pip install bandit",
            )

        target = paths if paths else [str(self.project_path)]

        try:
            result = run_python_module(
                "bandit",
                ["-r", "-f", "json"] + target,
                cwd=str(self.project_path),
            )

            issues = self.parse_output(result.stdout)

            return AnalyzerResult(
                tool=self.name,
                success=True,
                issues=issues,
                raw_output=result.stdout,
                summary={
                    "total_issues": len(issues),
                    "high_severity": sum(
                        1 for i in issues if i.severity == Severity.SECURITY
                    ),
                },
            )
        except Exception as e:
            return AnalyzerResult(
                tool=self.name,
                success=False,
                error=str(e),
            )

    def parse_output(self, output: str) -> list[Issue]:
        if not output.strip():
            return []

        try:
            data = json.loads(output)
        except json.JSONDecodeError:
            return []

        issues = []
        for item in data.get("results", []):
            # Map Bandit severity to our severity
            bandit_severity = item.get("issue_severity", "").upper()
            if bandit_severity == "HIGH":
                severity = Severity.SECURITY
            elif bandit_severity == "MEDIUM":
                severity = Severity.WARNING
            else:
                severity = Severity.INFO

            issues.append(
                Issue(
                    file=item.get("filename", ""),
                    line=item.get("line_number", 0),
                    column=0,
                    code=item.get("test_id", ""),
                    message=f"{item.get('issue_text', '')} (Confidence: {item.get('issue_confidence', '')})",
                    severity=severity,
                    tool=self.name,
                )
            )

        return issues


class VultureAnalyzer(AnalyzerBase):
    """
    Vulture - Dead code detector.

    Finds:
    - Unused functions
    - Unused classes
    - Unused variables
    - Unused imports
    """

    name = "vulture"
    description = "Dead code detector"

    def is_available(self) -> bool:
        try:
            result = run_python_module("vulture", ["--version"])
            return result.returncode == 0
        except Exception:
            return False

    def run(self, paths: Optional[list[str]] = None) -> AnalyzerResult:
        if not self.is_available():
            return AnalyzerResult(
                tool=self.name,
                success=False,
                error="Vulture is not installed. Run: pip install vulture",
            )

        target = paths if paths else [str(self.project_path)]

        try:
            result = run_python_module(
                "vulture",
                ["--min-confidence", "80"] + target,
                cwd=str(self.project_path),
            )

            issues = self.parse_output(result.stdout)

            return AnalyzerResult(
                tool=self.name,
                success=True,
                issues=issues,
                raw_output=result.stdout,
                summary={
                    "total_unused": len(issues),
                },
            )
        except Exception as e:
            return AnalyzerResult(
                tool=self.name,
                success=False,
                error=str(e),
            )

    def parse_output(self, output: str) -> list[Issue]:
        issues = []
        # Vulture output format: file.py:line: message (confidence%)
        pattern = r"(.+):(\d+): (.+) \((\d+)% confidence\)"

        for line in output.strip().split("\n"):
            if not line.strip():
                continue

            match = re.match(pattern, line)
            if match:
                issues.append(
                    Issue(
                        file=match.group(1),
                        line=int(match.group(2)),
                        column=0,
                        code="DEAD",
                        message=f"{match.group(3)} ({match.group(4)}% confidence)",
                        severity=Severity.WARNING,
                        tool=self.name,
                    )
                )

        return issues


class RadonAnalyzer(AnalyzerBase):
    """
    Radon - Complexity analyzer.

    Calculates:
    - Cyclomatic complexity (A-F scale)
    - Maintainability index
    - Raw metrics (LOC, comments, etc.)
    """

    name = "radon"
    description = "Cyclomatic complexity analyzer"

    # Complexity thresholds
    COMPLEXITY_GRADES = {
        "A": "Simple (1-5)",
        "B": "Low (6-10)",
        "C": "Moderate (11-20)",
        "D": "High (21-30)",
        "E": "Very High (31-40)",
        "F": "Complex (41+)",
    }

    def is_available(self) -> bool:
        try:
            result = run_python_module("radon", ["--version"])
            return result.returncode == 0
        except Exception:
            return False

    def run(self, paths: Optional[list[str]] = None) -> AnalyzerResult:
        if not self.is_available():
            return AnalyzerResult(
                tool=self.name,
                success=False,
                error="Radon is not installed. Run: pip install radon",
            )

        target = paths if paths else [str(self.project_path)]

        try:
            # Run radon cc (cyclomatic complexity) with JSON output
            result = run_python_module(
                "radon",
                ["cc", "-j", "-a"] + target,
                cwd=str(self.project_path),
            )

            issues = self.parse_output(result.stdout)

            return AnalyzerResult(
                tool=self.name,
                success=True,
                issues=issues,
                raw_output=result.stdout,
                summary={
                    "total_functions": len(issues),
                    "complex_functions": sum(
                        1 for i in issues if i.code in ("D", "E", "F")
                    ),
                },
            )
        except Exception as e:
            return AnalyzerResult(
                tool=self.name,
                success=False,
                error=str(e),
            )

    def parse_output(self, output: str) -> list[Issue]:
        if not output.strip():
            return []

        try:
            data = json.loads(output)
        except json.JSONDecodeError:
            return []

        issues = []
        for filename, functions in data.items():
            if filename == "error":
                continue

            for func in functions:
                rank = func.get("rank", "A")

                # Only report C or worse
                if rank not in ("C", "D", "E", "F"):
                    continue

                # Map rank to severity
                if rank in ("E", "F"):
                    severity = Severity.ERROR
                elif rank == "D":
                    severity = Severity.WARNING
                else:
                    severity = Severity.INFO

                issues.append(
                    Issue(
                        file=filename,
                        line=func.get("lineno", 0),
                        column=0,
                        code=rank,
                        message=f"{func.get('type', '')} '{func.get('name', '')}' has complexity {func.get('complexity', 0)} (Grade {rank}: {self.COMPLEXITY_GRADES.get(rank, '')})",
                        severity=severity,
                        tool=self.name,
                    )
                )

        return issues


@dataclass
class FullAnalysisResult:
    """Result from running all analyzers."""
    results: dict[str, AnalyzerResult]
    total_issues: int
    has_errors: bool
    has_security_issues: bool

    def format_report(self) -> str:
        """Format a complete analysis report."""
        lines = [
            "=" * 70,
            "SPINEHUB CODE ANALYSIS REPORT",
            "=" * 70,
            "",
            f"Total Issues: {self.total_issues}",
            f"Has Errors: {'YES' if self.has_errors else 'NO'}",
            f"Security Issues: {'YES' if self.has_security_issues else 'NO'}",
            "",
        ]

        for tool_name, result in self.results.items():
            lines.append("-" * 70)
            lines.append(f"{tool_name.upper()}")
            lines.append("-" * 70)

            if not result.success:
                lines.append(f"  ERROR: {result.error}")
                continue

            lines.append(f"  Issues: {result.issue_count}")

            if result.issues:
                lines.append("  Top Issues:")
                for issue in result.issues[:5]:
                    lines.append(
                        f"    [{issue.severity.value}] {issue.file}:{issue.line} - {issue.message[:60]}"
                    )
                if len(result.issues) > 5:
                    lines.append(f"    ... and {len(result.issues) - 5} more")

            lines.append("")

        lines.append("=" * 70)
        return "\n".join(lines)


class CodeAnalyzer:
    """
    Unified code analyzer that orchestrates all analysis tools.

    Usage:
        analyzer = CodeAnalyzer("/path/to/project")
        result = analyzer.run_all()
        print(result.format_report())
    """

    def __init__(self, project_path: str):
        self.project_path = Path(project_path)
        self.analyzers = {
            "ruff": RuffAnalyzer(project_path),
            "bandit": BanditAnalyzer(project_path),
            "vulture": VultureAnalyzer(project_path),
            "radon": RadonAnalyzer(project_path),
        }

    def check_tools(self) -> dict[str, bool]:
        """Check which tools are available."""
        return {name: analyzer.is_available() for name, analyzer in self.analyzers.items()}

    def run_all(self, paths: Optional[list[str]] = None) -> dict[str, AnalyzerResult]:
        """Run all available analyzers."""
        results = {}

        for name, analyzer in self.analyzers.items():
            results[name] = analyzer.run(paths)

        return results

    def run_single(
        self, tool: str, paths: Optional[list[str]] = None
    ) -> AnalyzerResult:
        """Run a single analyzer."""
        if tool not in self.analyzers:
            return AnalyzerResult(
                tool=tool,
                success=False,
                error=f"Unknown tool: {tool}. Available: {list(self.analyzers.keys())}",
            )

        return self.analyzers[tool].run(paths)

File: src/spinehub/analyzers/code-analyzer.ts
/**
 * Code Analyzer - TypeScript wrapper for Python code analysis tools
 *
 * Provides access to:
 * - Ruff (fast Python linter)
 * - Bandit (security linter)
 * - Vulture (dead code detection)
 * - Radon (complexity analysis)
 */

import { getPythonBridge, BridgeResponse } from '../bridge';
import {
  ToolStatus,
  AnalyzerResult,
  Issue,
  RunAllAnalyzersResult,
} from '../types';

export type AnalyzerTool = 'ruff' | 'bandit' | 'vulture' | 'radon';

export interface CodeAnalyzerOptions {
  excludePaths?: string[];
  autoFix?: boolean;
}

/**
 * CodeAnalyzer - High-level interface for code analysis
 */
export class CodeAnalyzer {
  private options: CodeAnalyzerOptions;

  constructor(options: CodeAnalyzerOptions = {}) {
    this.options = {
      excludePaths: options.excludePaths || [],
      autoFix: options.autoFix || false,
    };
  }

  /**
   * Check which analysis tools are installed
   */
  async checkTools(): Promise<ToolStatus> {
    const bridge = getPythonBridge();
    const response = await bridge.call<ToolStatus>('analyzers.check_tools', {});

    if (!response.success) {
      console.error('Failed to check tools:', response.error);
      return {
        ruff: false,
        bandit: false,
        vulture: false,
        radon: false,
      };
    }

    return response.data!;
  }

  /**
   * Run all available analyzers on specified paths
   */
  async runAll(paths?: string[]): Promise<RunAllAnalyzersResult> {
    const bridge = getPythonBridge();
    const response = await bridge.call<RunAllAnalyzersResult>('analyzers.run_all', {
      paths: paths || ['.'],
      exclude: this.options.excludePaths,
    });

    if (!response.success) {
      console.error('Analysis failed:', response.error);
      return {
        tools_status: { ruff: false, bandit: false, vulture: false, radon: false },
        results: {},
        total_issues: 0,
        by_severity: {},
      };
    }

    return response.data!;
  }

  /**
   * Run a single analyzer
   */
  async runSingle(tool: AnalyzerTool, paths?: string[]): Promise<AnalyzerResult> {
    const bridge = getPythonBridge();
    const response = await bridge.call<AnalyzerResult>('analyzers.run_single', {
      tool,
      paths: paths || ['.'],
      exclude: this.options.excludePaths,
    });

    if (!response.success) {
      return {
        tool,
        success: false,
        issues: [],
        summary: {},
        error: response.error,
      };
    }

    return response.data!;
  }

  /**
   * Run Ruff linter
   */
  async runRuff(paths?: string[]): Promise<AnalyzerResult> {
    return this.runSingle('ruff', paths);
  }

  /**
   * Run Bandit security scanner
   */
  async runBandit(paths?: string[]): Promise<AnalyzerResult> {
    return this.runSingle('bandit', paths);
  }

  /**
   * Run Vulture dead code detector
   */
  async runVulture(paths?: string[]): Promise<AnalyzerResult> {
    return this.runSingle('vulture', paths);
  }

  /**
   * Run Radon complexity analyzer
   */
  async runRadon(paths?: string[]): Promise<AnalyzerResult> {
    return this.runSingle('radon', paths);
  }

  /**
   * Get issues filtered by severity
   */
  filterBySeverity(results: RunAllAnalyzersResult, severity: Issue['severity']): Issue[] {
    const allIssues: Issue[] = [];

    for (const result of Object.values(results.results)) {
      allIssues.push(...result.issues.filter(i => i.severity === severity));
    }

    return allIssues;
  }

  /**
   * Get security issues only (from Bandit)
   */
  getSecurityIssues(results: RunAllAnalyzersResult): Issue[] {
    return this.filterBySeverity(results, 'security');
  }

  /**
   * Format results as readable report
   */
  formatReport(results: RunAllAnalyzersResult): string {
    const lines: string[] = [];

    lines.push('=' .repeat(60));
    lines.push('         CODE ANALYSIS REPORT');
    lines.push('=' .repeat(60));
    lines.push('');

    // Tool status
    lines.push('[TOOLS]');
    for (const [tool, available] of Object.entries(results.tools_status)) {
      const status = available ? '[OK]' : '[--]';
      lines.push(`  ${status} ${tool}`);
    }
    lines.push('');

    // Summary
    lines.push('[SUMMARY]');
    lines.push(`  Total Issues: ${results.total_issues}`);
    for (const [severity, count] of Object.entries(results.by_severity)) {
      lines.push(`  ${severity}: ${count}`);
    }
    lines.push('');

    // Issues by tool
    for (const [tool, result] of Object.entries(results.results)) {
      if (result.issues.length === 0) continue;

      lines.push(`[${tool.toUpperCase()}] (${result.issues.length} issues)`);
      for (const issue of result.issues.slice(0, 10)) {
        lines.push(`  ${issue.file}:${issue.line} - ${issue.code}: ${issue.message}`);
      }
      if (result.issues.length > 10) {
        lines.push(`  ... and ${result.issues.length - 10} more`);
      }
      lines.push('');
    }

    lines.push('=' .repeat(60));

    return lines.join('\n');
  }
}

/**
 * Quick analysis function for CLI usage
 */
export async function analyzeCode(paths?: string[]): Promise<RunAllAnalyzersResult> {
  const analyzer = new CodeAnalyzer();
  return analyzer.runAll(paths);
}

/**
 * Check if analysis tools are available
 */
export async function checkAnalysisTools(): Promise<ToolStatus> {
  const analyzer = new CodeAnalyzer();
  return analyzer.checkTools();
}
